### DISCUSSION TITLE
Validate List[str] using regex

### DISCUSSION THREAD

**Discussion by marius-benthin** [fastapi/fastapi#7806] [Questions] [answered]
Hello :)

How do I validate a list of strings that are sent via POST parameters?

I have this example ...
```python
@app.post("/entity/list")
def query_entity_list(entities: List[str] = Query(None, regex="^[a-z]{5}$")):
   return entities
```
... and the API function is called by ...
```
curl -X POST "http://localhost/entity/list/?entities=valid&entities=inv4lid
```
Is it possible to apply the regex on each item in the entities list?
Currently, the string validation does not working for me.

Thanks in advance!

**Comment by sm-Fifteen** [comment]
You would want to have a specific string type with that contraint. Pydantic has a `ConstrainedStr`/`constr` type spcifically for that purpose:

```py
from pydantic import constr

EntryCode = constr(regex="^[a-z]{5}$")

@app.post("/entity/list")
def query_entity_list(entities: List[EntryCode] = Query(None)):
   return entities
```

**Reply by YuriiMotov** [reply, marked as answer]
Modern alternative with better typing:

```python
from typing import Annotated, TypeAlias
from fastapi import FastAPI, Query
from pydantic import Field

app = FastAPI()

EntryCode: TypeAlias = Annotated[str, Field(pattern="^[a-z]{5}$")]

@app.post("/entity/list")
def query_entity_list(entities: Annotated[list[EntryCode], Query()]):
    return entities
```

**Comment by marius-benthin** [comment]
Thank you very much @sm-Fifteen! It works now :)

**Comment by tiangolo** [comment]
Thanks for the help here @sm-Fifteen ! :cake: 

Thanks @marius-benthin for reporting back and closing the issue :+1:

**Comment by dwbelliston** [comment]
Is there a way to preserve the regex annotation?

when i add constr the docs show just `str`

```
    use_filters: Optional[List[constr(regex=models.SEGMENT_PARAM_REGEX)]] = Query(
        None,
        alias="useFilters",
        title="Use Filters",
        # regex=models.SEGMENT_PARAM_REGEX,
        description="Filter results using Field:Operator:Value",
    ),
```    


![image](https://user-images.githubusercontent.com/11450118/109264009-804f3480-77c1-11eb-9a9f-c110aa16b436.png)


but with the regex on Query it shows up in the docs

```
    use_filters: Optional[List[str]] = Query(
        None,
        alias="useFilters",
        title="Use Filters",
        regex=models.SEGMENT_PARAM_REGEX,
        description="Filter results using Field:Operator:Value",
    ),
```

![image](https://user-images.githubusercontent.com/11450118/109264097-ad034c00-77c1-11eb-8bbf-e9d30eba4ae0.png)

**Reply by YuriiMotov** [reply]
With the following code it produces correct openapi schema:
```python
from typing import Annotated, TypeAlias
from fastapi import FastAPI, Query
from pydantic import Field

app = FastAPI()

EntryCode: TypeAlias = Annotated[str, Field(pattern="^[a-z]{5}$")]

@app.post("/entity/list")
def query_entity_list(entities: Annotated[list[EntryCode], Query()]):
    return entities
```

```json
        "parameters": [
          {
            "name": "entities",
            "in": "query",
            "required": true,
            "schema": {
              "type": "array",
              "items": {
                "type": "string",
                "pattern": "^[a-z]{5}$"
              },
              "title": "Entities"
            }
          }
        ],
```

Value validation works in Swagger UI (try sending incorrect string). Swagger just doen't display this pattern.

With the regex on Query it produced incorrect schema (pattern was added to both, array and item).
In current version (Pydantic 2.11.7) it doesn't take effect and produces the same schema as without it
